presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ImplementIntersectAsUnion.java

```diff
@@ -0,0 +1,225 @@
+import com.facebook.presto.sql.tree.NullLiteral;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.QualifiedNameReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.metadata.FunctionKind.AGGREGATE;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.facebook.presto.sql.tree.ComparisonExpression.Type.GREATER_THAN_OR_EQUAL;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Converts INTERSECT query into UNION ALL..GROUP BY...WHERE
+ * Eg:  SELECT a FROM foo INTERSECT SELECT x FROM bar
+ * <p>
+ * =>
+ * <p>
+ * SELECT a
+ * FROM
+ * (SELECT a,
+ * COUNT(foo_marker) AS foo_cnt,
+ * COUNT(bar_marker) AS bar_cnt
+ * FROM
+ * (
+ * SELECT a, true as foo_marker, null as bar_marker FROM foo
+ * UNION ALL
+ * SELECT x, null as foo_marker, true as bar_marker FROM bar
+ * ) T1
+ * GROUP BY a) T2
+ * WHERE foo_cnt >= 1 AND bar_cnt >= 1;
+ */
+public class ImplementIntersectAsUnion
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(symbolAllocator, "symbolAllocator is null");
+        requireNonNull(idAllocator, "idAllocator is null");
+
+        return SimplePlanRewriter.rewriteWith(new Rewriter(idAllocator, symbolAllocator), plan);
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private static final String INTERSECT_MARKER = "intersect_marker";
+        private final PlanNodeIdAllocator idAllocator;
+        private final SymbolAllocator symbolAllocator;
+
+        private Rewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitIntersect(IntersectNode node, RewriteContext<Void> rewriteContext)
+        {
+            List<PlanNode> sources = node.getSources();
```

Comment:
`sources` need to be processed recursively (by calling `rewriteContext.rewrite(...)`). Otherwise, if there's an IntersectNode nested in a subquery, it won't be translated. You should add a unit test that covers this case, too.